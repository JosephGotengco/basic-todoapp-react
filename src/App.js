import React from "react";
import { Route, Switch } from "react-router-dom";
import About from "./components/pages/About";
import Home from "./components/pages/Home";
import Api from "./components/pages/Api";
import Navbar from "./components/Navbar";
import Error from "./components/Error";
import ThemeSwitcher from "./components/ThemeSwitcher";

// Route component tells which component to display based on the path
function App() {
    return (
        <div className="todoapp stack-large">
            <div className="nav-group">
                <Navbar />
                <ThemeSwitcher />
            </div>
            <Switch>
                <Route path="/" component={Home} exact />
                <Route path="/about" component={About} />
                <Route path="/api" component={Api} />
                <Route component={Error} />
            </Switch>
        </div>
    );
}

export default App;
