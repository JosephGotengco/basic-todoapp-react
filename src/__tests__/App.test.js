import React from "react";
import renderer from "react-test-renderer";
import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import App from "../App";
import Todo from "../components/Todo";

describe("todoapp tests", () => {
    const NO_TASKS = [];
    const TASKS = [
        { id: "todo-0", name: "Unit", completed: true, deleteTask: (id) => {} },
        { id: "todo-1", name: "Test", completed: true },
        { id: "todo-2", name: "Everything", completed: false },
    ];
    const FILTERS = [
        { name: "All", pressed: true },
        { name: "Active", pressed: false },
        { name: "Completed", pressed: false },
    ];

    describe("App.js unit tests", () => {
        it("filter buttons should be in the correct state when initialized", () => {
            const component = renderer.create(<App tasks={TASKS} />);
            const testInstance = component.toJSON();
            const filterButtons = testInstance.children[1].children;
            filterButtons.forEach((button, i) => {
                expect(button.props["aria-pressed"]).toBe(FILTERS[i].pressed);
                expect(button.children[1].children[0]).toBe(FILTERS[i].name);
            });
        });
    });

    describe("Form.js unit tests", () => {
        it("should be able to change value of new todo input", () => {
            render(<App tasks={NO_TASKS} />);

            userEvent.type(screen.getByTestId("new-todo-input"), "new todo");
            const input = screen.getByTestId("new-todo-input");
            expect(input.value).toBe("new todo");
        });

        it("should be able to change new todo input value", () => {
            render(<App tasks={NO_TASKS} />);
            userEvent.type(screen.getByTestId("new-todo-input"), "new todo");
            const input = screen.getByTestId("new-todo-input");
            expect(input.value).toBe("new todo");
        });
    });

    describe("Todo.js unit tests", function () {
        it("should be able to render todo edit form", () => {
            const { getByTestId } = render(<Todo tasks={TASKS[0]} />);
            const checkbox = getByTestId("todo-view");
            expect(checkbox.constructor.name).toBe("HTMLDivElement");
            userEvent.click(screen.getByTestId("todo-view-edit-btn"));
            // verify todo view form isn't snowhing anymore
            const editBtn = screen.queryByText("edit");
            expect(editBtn).toBeNull();
            // verify todo edit form is showing
            getByTestId("todo-edit-view");
        });
    });

    describe("snapshot tests", () => {
        it("should match snapshot with tasks when rendered", () => {
            const tree = renderer.create(<App tasks={TASKS} />).toJSON();
            expect(tree).toMatchSnapshot();
        });
    });
});
