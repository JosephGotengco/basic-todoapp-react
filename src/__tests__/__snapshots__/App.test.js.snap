// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todoapp tests snapshot tests should match snapshot with tasks when rendered 1`] = `
<div
  className="todoapp stack-large"
>
  <form
    onSubmit={[Function]}
  >
    <h2
      className="label-wrapper"
    >
      <label
        className="label__lg"
        htmlFor="new-todo-input"
      >
        What needs to be done?
      </label>
    </h2>
    <input
      autoComplete="off"
      className="input input__lg"
      data-testid="new-todo-input"
      id="new-todo-input"
      name="text"
      onChange={[Function]}
      type="text"
      value=""
    />
    <button
      className="btn btn__primary btn__lg"
      data-testid="new-todo-button"
      type="submit"
    >
      Add
    </button>
  </form>
  <div
    className="filters btn-group stack-exception"
  >
    <button
      aria-pressed={true}
      className="btn toggle-btn"
      onClick={[Function]}
      type="button"
    >
      <span
        className="visually-hidden"
      >
        Show 
      </span>
      <span>
        All
      </span>
      <span
        className="visually-hidden"
      >
         tasks
      </span>
    </button>
    <button
      aria-pressed={false}
      className="btn toggle-btn"
      onClick={[Function]}
      type="button"
    >
      <span
        className="visually-hidden"
      >
        Show 
      </span>
      <span>
        Active
      </span>
      <span
        className="visually-hidden"
      >
         tasks
      </span>
    </button>
    <button
      aria-pressed={false}
      className="btn toggle-btn"
      onClick={[Function]}
      type="button"
    >
      <span
        className="visually-hidden"
      >
        Show 
      </span>
      <span>
        Completed
      </span>
      <span
        className="visually-hidden"
      >
         tasks
      </span>
    </button>
  </div>
  <h2
    id="list-heading"
  >
    3 tasks remaining
  </h2>
  <ul
    aria-labelledby="list-heading"
    className="todo-list stack-large stack-exception"
    role="list"
  >
    <li
      className="todo"
    >
      <div
        className="stack-small"
        data-testid="todo-view"
      >
        <div
          className="c-cb"
        >
          <input
            defaultChecked={true}
            id="todo-0"
            onChange={[Function]}
            type="checkbox"
          />
          <label
            className="todo-label"
            htmlFor="todo-0"
          >
            Unit
          </label>
        </div>
        <div
          className="btn-group"
        >
          <button
            className="btn"
            data-testid="todo-view-edit-btn"
            onClick={[Function]}
            type="button"
          >
            Edit 
            <span
              className="visually-hidden"
            >
              Unit
            </span>
          </button>
          <button
            className="btn btn__danger"
            onClick={[Function]}
            type="button"
          >
            Delete 
            <span
              className="visually-hidden"
            >
              Unit
            </span>
          </button>
        </div>
      </div>
    </li>
    <li
      className="todo"
    >
      <div
        className="stack-small"
        data-testid="todo-view"
      >
        <div
          className="c-cb"
        >
          <input
            defaultChecked={true}
            id="todo-1"
            onChange={[Function]}
            type="checkbox"
          />
          <label
            className="todo-label"
            htmlFor="todo-1"
          >
            Test
          </label>
        </div>
        <div
          className="btn-group"
        >
          <button
            className="btn"
            data-testid="todo-view-edit-btn"
            onClick={[Function]}
            type="button"
          >
            Edit 
            <span
              className="visually-hidden"
            >
              Test
            </span>
          </button>
          <button
            className="btn btn__danger"
            onClick={[Function]}
            type="button"
          >
            Delete 
            <span
              className="visually-hidden"
            >
              Test
            </span>
          </button>
        </div>
      </div>
    </li>
    <li
      className="todo"
    >
      <div
        className="stack-small"
        data-testid="todo-view"
      >
        <div
          className="c-cb"
        >
          <input
            defaultChecked={false}
            id="todo-2"
            onChange={[Function]}
            type="checkbox"
          />
          <label
            className="todo-label"
            htmlFor="todo-2"
          >
            Everything
          </label>
        </div>
        <div
          className="btn-group"
        >
          <button
            className="btn"
            data-testid="todo-view-edit-btn"
            onClick={[Function]}
            type="button"
          >
            Edit 
            <span
              className="visually-hidden"
            >
              Everything
            </span>
          </button>
          <button
            className="btn btn__danger"
            onClick={[Function]}
            type="button"
          >
            Delete 
            <span
              className="visually-hidden"
            >
              Everything
            </span>
          </button>
        </div>
      </div>
    </li>
  </ul>
</div>
`;
